agent/agentcontainers/containers_dockercli.go

```diff
@@ -31,6 +33,172 @@ func NewDocker(execer agentexec.Execer) Lister {
+		return nil, xerrors.Errorf("get container user: scan /etc/passwd: %w", err)
+	}
+	if foundLine == "" {
+		return nil, xerrors.Errorf("get container user: no matching entry for %q found in /etc/passwd", containerUser)
+	}
+
+	// Parse the output of /etc/passwd. It looks like this:
+	// postgres:x:999:999::/var/lib/postgresql:/bin/bash
+	passwdFields := strings.Split(foundLine, ":")
+	if len(passwdFields) < 7 {
+		return nil, xerrors.Errorf("get container user: invalid line in /etc/passwd: %q", foundLine)
+	}
+
+	// The fourth entry in /etc/passwd contains GECOS information, which is a
+	// comma-separated list of fields. The first field is the user's full name.
+	gecos := strings.Split(passwdFields[4], ",")
+	fullName := ""
+	if len(gecos) > 1 {
+		fullName = gecos[0]
+	}
+
+	dei.user = &user.User{
+		Gid:      passwdFields[3],
+		HomeDir:  passwdFields[5],
+		Name:     fullName,
+		Uid:      passwdFields[2],
+		Username: containerUser,
+	}
+	dei.userShell = passwdFields[6]
+
+	// Finally, get the environment of the container.
+	stdoutBuf.Reset()
+	stderrBuf.Reset()
+	cmd, args = WrapDockerExec(container, containerUser)("env")
+	execCmd = execer.CommandContext(ctx, cmd, args...)
+	execCmd.Stdout = &stdoutBuf
+	execCmd.Stderr = &stderrBuf
+	if err := execCmd.Run(); err != nil {
+		return nil, xerrors.Errorf("get container environment: run env: %w stderr: %q", err, strings.TrimSpace(stderrBuf.String()))
+	}
+
+	scanner = bufio.NewScanner(&stdoutBuf)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		dei.env = append(dei.env, line)
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, xerrors.Errorf("get container environment: scan env output: %w", err)
+	}
+
+	return &dei, nil
+}
+
+func (dei *ContainerEnvInfoer) CurrentUser() (*user.User, error) {
+	// Clone the user so that the caller can't modify it
+	u := &user.User{
+		Gid:      dei.user.Gid,
+		HomeDir:  dei.user.HomeDir,
+		Name:     dei.user.Name,
+		Uid:      dei.user.Uid,
+		Username: dei.user.Username,
+	}
+	return u, nil
+}
+
+func (dei *ContainerEnvInfoer) Environ() []string {
+	// Return a clone of the environment so that the caller can't modify it
+	return slices.Clone(dei.env)
+}
+
+func (dei *ContainerEnvInfoer) UserHomeDir() (string, error) {
+	return dei.user.HomeDir, nil
+}
+
+func (dei *ContainerEnvInfoer) UserShell(string) (string, error) {
+	return dei.userShell, nil
+}
+
+func (dei *ContainerEnvInfoer) ModifyCommand(cmd string, args ...string) (string, []string) {
+	return WrapDockerExecPTY(dei.container, dei.user.Username)(cmd, args...)
+}
+
+// WrapFn is a function that wraps a command and its arguments with another command and arguments.
+type WrapFn func(cmd string, args ...string) (string, []string)
+
+// WrapDockerExec returns a WrapFn that wraps the given command and arguments
+// with a docker exec command that runs as the given user in the given container.
+func WrapDockerExec(containerName, userName string) WrapFn {
```

Comment:
These feel familiar, is it duplication or am I hallucinating? 😄

Reply:
You probably remember them from https://github.com/coder/coder/pull/16552 -- I ended up throwing all my changes in here and picking them out bit by bit.